presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBuilder.java

```diff
@@ -0,0 +1,385 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+
+public class ArrayBuilder
+        extends AbstractVariableWidthBlock
+        implements BlockBuilder
+{
+    private final BlockBuilderStatus builderStatus;
+    private final VariableWidthBlockBuilder builder;
+    private final List<Type> parameterTypes;
+    private final Iterator<Type> typeIterator;
+    private final boolean valueHolder;
+
+    private Slice finalValue;
+
+    public static ArrayBuilder builder(Type parameterType, Type ... parameterTypes)
+    {
+        return new ArrayBuilder(toList(parameterType, parameterTypes));
+    }
+
+    public static ArrayBuilder builder(List<Type> parameterTypes)
+    {
+        return new ArrayBuilder(parameterTypes);
+    }
+
+    public static ArrayBuilder builder(int expectedSize, Type parameterType, Type... parameterTypes)
+    {
+        return new ArrayBuilder(expectedSize, false, toList(parameterType, parameterTypes));
+    }
+
+    public static ArrayBuilder builder(int expectedSize, List<Type> parameterTypes)
+    {
+        return new ArrayBuilder(expectedSize, false, parameterTypes);
+    }
+
+    public static ArrayBuilder valueHolder()
```

Comment:
Lets make the "value holder" a separate class (can be a subclass), instead of adding it to ArrayBuilder. Also, add a check that `writeTo` is only called once, as otherwise the second write would overwrite the first one